File: packages/dataframe/rust/stats/distributions/chi_squared.test.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { callR, printResults, runComparison } from "./test-helpers.ts";

async function main() {
  console.log("Testing R vs Rust chi-squared functions...\n");

  const testCases = [
    { func: "dchisq", distribution: "chi_squared", args: [1, 5, 0] },
    { func: "pchisq", distribution: "chi_squared", args: [1, 5, 1, 0] },
    { func: "qchisq", distribution: "chi_squared", args: [0.5, 5, 1, 0] },
  ];

  const results = await runComparison(testCases);
  printResults(results);

  // Test rchisq separately since it's not available in Rust interface
  console.log("\nR-only tests:");
  try {
    const rchisqResult = await callR("rchisq", "chi_squared", 1, 5);
    console.log(`rchisq(5) = ${rchisqResult}`);
  } catch (error) {
    console.log(`rchisq: Error - ${(error as Error).message}`);
  }
}

if (import.meta.main) {
  main().catch(console.error);
}

File: packages/dataframe/rust/stats/distributions/chi_squared.test.R
#!/usr/bin/env Rscript

# Chi-squared distribution test script
# Usage: Rscript chi_squared.test.R <function> <x_or_p> <df> [lower_tail] [log_p]

source("test-helpers.R")

# Define chi-squared distribution functions
chi_squared_density <- function(x, df, log = FALSE) {
  dchisq(x, df, log = log)
}

chi_squared_cumulative <- function(x, df, lower.tail = TRUE, log.p = FALSE) {
  pchisq(x, df, lower.tail = lower.tail, log.p = log.p)
}

chi_squared_quantile <- function(p, df, lower.tail = TRUE, log.p = FALSE) {
  qchisq(p, df, lower.tail = lower.tail, log.p = log.p)
}

chi_squared_random <- function(n, df) {
  rchisq(n, df)
}

# Run the test
run_distribution_test(chi_squared_density, chi_squared_cumulative, chi_squared_quantile, chi_squared_random)
